cranelift/codegen/src/egraph.rs

```diff
@@ -0,0 +1,412 @@
+                        mem_state,
+                        srcloc,
+                    }
+                } else if side_effect {
+                    self.stats.node_created += 1;
+                    self.stats.node_inst += 1;
+                    Node::Inst {
+                        op,
+                        inst,
+                        args,
+                        types,
+                        srcloc,
+                        loop_level,
+                    }
+                } else {
+                    self.stats.node_created += 1;
+                    self.stats.node_pure += 1;
+                    Node::Pure { op, args, types }
+                };
+                let dedup_needed = self.node_ctx.needs_dedup(&node);
+
+                let id = self.egraph.add(node, &mut self.node_ctx);
+
+                if dedup_needed {
+                    self.stats.node_dedup_query += 1;
+                    match id {
+                        NewOrExisting::New(_) => {
+                            self.stats.node_dedup_miss += 1;
+                        }
+                        NewOrExisting::Existing(_) => {
+                            self.stats.node_dedup_hit += 1;
+                        }
+                    }
+                }
+
+                if opcode == Opcode::Store {
+                    let store_data_ty = func.dfg.value_type(func.dfg.inst_args(inst)[0]);
+                    self.store_nodes.insert(inst, (store_data_ty, id.get()));
+                    self.stats.store_map_insert += 1;
+                }
+
+                let id = match (side_effect, mem_state, id) {
+                    (_, Some(..), NewOrExisting::New(id)) => {
+                        // Loads: do store-to-load forwarding, and
+                        // otherwise add to side-effecting roots.
+                        let opt_id = crate::opts::store_to_load(id, self);
+                        trace!("store_to_load: {} -> {}", id, opt_id);
+                        if opt_id == id {
+                            self.side_effect_ids.push(id);
+                            self.stats.side_effect_nodes += 1;
+                        }
+                        opt_id
+                    }
+                    (true, _, id) => {
+                        let id = id.get();
+                        self.side_effect_ids.push(id);
+                        self.stats.side_effect_nodes += 1;
+                        id
+                    }
+                    (false, _, NewOrExisting::New(id)) => {
+                        // Apply all optimization rules immediately; the
+                        // aegraph (acyclic egraph) works best when we do
+                        // this so all uses pick up the eclass with all
+                        // possible enodes.
+                        crate::opts::optimize_eclass(id, self)
+                    }
+                    (false, _, NewOrExisting::Existing(id)) => id,
+                };
```

Comment:
Unfortunately I think this will be a meaningful performance regression -- I remember specifically restricting store-to-load logic to run only on loads while tuning.

Happy to add comments here to clarify as needed -- any thoughts (while keeping the structure the same)?

Reply:
OK, talked with @jameysharp offline about this one and I think we can get around the perf regression with another conditional; and the rest of this is actually nice and clean and I like it better. So I'll accept this then tweak it a bit.